### PR TITLE
[WIP] Fix: protect Observable's listener map from changing when ranging

### PR DESCRIPTION
**Pull request by comzyh**

### CODE REVIEW

File: common/observable/observable.go [hunk 1]
```diff
@@ -6,19 +6,22 @@ import (
...
 type Observable struct {
-	iterable Iterable
-	listener *sync.Map
-	done     bool
-	doneLock sync.RWMutex
+	iterable     Iterable
+	listener     *sync.Map
+	done         bool
+	doneLock     sync.RWMutex
+	listenerlock sync.Mutex
```
Comment (tslling) [inline common/observable/observable.go]: camel case: ```listenerLock``` :)

**Comment by Dreamacro** [comment]
maybe refactor this pkg is the best choice

**Comment by comzyh** [comment]
> maybe refactor this pkg is the best choice

Sure, I'm thinking about it.

**Comment by Dreamacro** [comment]
@comzyh cc

**Comment by comzyh** [comment]
LGTM, thanks!

**Pull request merged**